### PR TITLE
Cache mailer view

### PR DESCRIPTION
**Pull request by st0012**
Hi, I'm adding cache feature to `ActionMailer`'s templates. I think there two ways to accomplish this goal. 
First is moving all the caching feature from `ActionController` to `AbstractController` since both `ActionMailer` and `ActionController` are inherited from it.
The second one is porting caching features to `ActionMailer`, which I think will break the `DRY` rule.
However, I'm working on this with the second approach. Because I think after I reimplemented it in `ActionMailer`, I can find some reusable parts in both `ActionMailer` and `ActionController` and put them into somewhere like `ActiveSupport`?
And I know this could be a huge work while I am not every familiar with Rails' codebase. So I am hoping someone can discuss this with me or give me some advice. Any suggestion will be very helpful, thanks!

### CODE REVIEW

File: actionmailer/.byebug_history [hunk 1]
```diff
@@ -0,0 +1,8 @@
+exit
```
Comment (tadeusrox) [inline actionmailer/.byebug_history]: I think this file shouldn't be here

If you think it's usefull -> https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

Comment (st0012) [inline actionmailer/.byebug_history]: @tadeusrox  Thanks!

**Comment by st0012** [comment]
And I think the main difference between `ActionController` and `ActionMailer`'s caching implementation is how the cache key is generated and `ActionMailer` doesn't need to deal with requests.

**Comment by rafaelfranca** [comment]
> I can find some reusable parts in both ActionMailer and ActionController and put them into somewhere like ActiveSupport?

Yes. You can put in Action Dispatch like Abstract Controller already is.

**Comment by st0012** [comment]
Hi, I'm wondering if I need to copy all the caching test from `ActionController`, and change them to the `ActionMailer` version? I think most of them are testing the fragment cache itself, not testing `fragment cache in ActionController`. So make a copy of them in `ActionMailer` are basically unnecessary. What do you think? cc @rafaelfranca @arthurnn

**Comment by st0012** [comment]
I don't know why CI would fail. It passed in my local VM with command: `GEM=am,amo,as,av,aj bundle exec rake test`.

**Comment by st0012** [comment]
@arthurnn Hi, can you please give me some advice? thank you :smile:

**Comment by rafaelfranca** [comment]
> Hi, I'm wondering if I need to copy all the caching test from ActionController, and change them to the ActionMailer version? I think most of them are testing the fragment cache itself, not testing fragment cache in ActionController. So make a copy of them in ActionMailer are basically unnecessary. What do you think?

Yes, agree. It is better to have a fresh start in the test suite, but it is good to use the ActionController tests as inspiration.

File: actionmailer/lib/action_mailer/caching.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+require 'action_dispatch/caching'
+
+module ActionMailer
+  module Caching
+    extend ActiveSupport::Autoload
```
Comment (rafaelfranca) [inline actionmailer/lib/action_mailer/caching.rb]: you don't need this.

File: actionmailer/lib/action_mailer/caching.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'action_dispatch/caching'
+
+module ActionMailer
+  module Caching
```
Comment (rafaelfranca) [inline actionmailer/lib/action_mailer/caching.rb]: I think we can remove this module and just include its content directly in `ActionMailer::Base`

File: actionpack/lib/action_controller/caching.rb [hunk 4]
```diff
@@ -23,65 +22,24 @@ module ActionController
   #   config.action_controller.cache_store = :mem_cache_store, Memcached::Rails.new('localhost:11211')
   #   config.action_controller.cache_store = MyOwnStore.new('parameter')
   module Caching
-    extend ActiveSupport::Concern
     extend ActiveSupport::Autoload
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/caching.rb]: This line can be removed.

File: actionpack/lib/action_dispatch/caching.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      end
+    end
+
+    def view_cache_dependencies
+      self.class._view_cache_dependencies.map { |dep| instance_exec(&dep) }.compact
+    end
+
+    protected
+      # Convenience accessor.
+    def cache(key, options = {}, &block)
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/caching.rb]: The indentation here needs to be fixed

File: actionpack/lib/action_dispatch/caching.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        config.cache_store = ActiveSupport::Cache.lookup_store(store)
+      end
+
+      private
+        def cache_configured?
+          perform_caching && cache_store
+        end
+    end
+
+    include AbstractController::Helpers
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/caching.rb]: Why is this included here? It is being used here somewhere? I think it is fine to remove since both classes that include `AD::Caching` already include this module.

File: actionmailer/lib/action_mailer/caching.rb (hunk 3 above)
Comment (st0012) [inline actionmailer/lib/action_mailer/caching.rb]: Should we also move `ActionController::Caching` into `ActionController::Base`?

File: actionpack/lib/action_controller/caching.rb (hunk 4 above)
Comment (st0012) [inline actionpack/lib/action_controller/caching.rb]: Remove this line will cause `undefined method eager_load!` error in [here](https://github.com/rails/rails/blob/master/actionpack%2Flib%2Faction_controller.rb#L49).

**Comment by st0012** [comment]
Here's what I have done for now:
1. I move the `ActionController::Caching` to `ActionDispatch::Caching` and include it in `ActionMailer::Base` and `ActionController::Caching`.
2. I added some unit tests and functional test to make sure that `ActionMailer` has the caching ability.
3. Finally, I changed the environment template files to add `ActionMailer`'s caching configuration.

And I am wondering what I should do next? This is my first time contributing `Rails`, so I think I need some guide to finish this :cry: .

**Comment by st0012** [comment]
@rafaelfranca Hello ~ what can I do to merge this?

**Comment by st0012** [comment]
@rafaelfranca Thanks for merging!

**Comment by st0012** [comment]
@rafaelfranca BTW, when should/can I edit the documentations?

**Comment by rafaelfranca** [comment]
You can open a PR upgrading the documentation, and we really should.

**Comment by st0012** [comment]
@rafaelfranca ok, thanks!
